### ISSUE TITLE
Implement Request Timeout and Retry Mechanism for LLM Operations - T152

### ISSUE THREAD

**Issue by Krish3017** [gdg-charusat/PDF_QA_Bot#33]
### Team Number - T152
# Implement Request Timeout and Retry Mechanism for LLM Operations 

## 📋 Issue Type
- [x] Enhancement
- [x] Bug Fix
- [x] Performance Improvement

## 🎯 Priority
**Level:** 2 (Medium-High Complexity)  
**Labels:** `enhancement`, `bug`, `level-2`, `Sprintathon`, `reliability`

## 🔴 Problem Statement
No timeout handling exists for long-running LLM generation requests. If the model hangs or takes too long, requests never complete, blocking the service and creating poor user experience. This is a potential DoS vulnerability.

**Current Behavior:**
- Requests can hang indefinitely
- No timeout on `model.generate()`
- No timeout on axios calls to RAG service
- No retry logic for transient failures
- Frontend waits forever with spinning loader

## 💥 Impact
- **Resource Exhaustion:** Hanging requests consume server memory/CPU
- **User Experience:** Users wait indefinitely without feedback
- **Service Availability:** Blocked resources prevent new requests
- **DoS Vulnerability:** Malicious users can exhaust server resources
- **Production Readiness:** Unsuitable for production deployment
- **No Graceful Degradation:** System can't recover from transient errors

### RAG Service (rag-service/main.py)
**Lines 55-87** - No timeout in generation:
```python
def generate_response(prompt: str, max_new_tokens: int) -> str:
    # ... setup code ...
    with torch.no_grad():
        generated_ids = model.generate(
            **encoded,
            max_new_tokens=max_new_tokens,
            do_sample=False,
            pad_token_id=pad_token_id,
            # ❌ No timeout parameter!
        )
    # Can hang forever here
```

### Backend (server.js)
**Lines 38-65** - No timeout configuration:
```python
app.post("/ask", async (req, res) => {
    try {
        const response = await axios.post(
            "http://localhost:5000/ask",
            { question: question.trim() }
            // ❌ No timeout option!
        );
    } catch (err) {
        // Generic error handling
    }
});
```

### 1. **Add LLM Generation Timeout**
```python
# In generate_response()
import signal
from contextlib import contextmanager

@contextmanager
def timeout(seconds):
    def timeout_handler(signum, frame):
        raise TimeoutError("Generation timed out")
    signal.signal(signal.SIGALRM, timeout_handler)
    signal.alarm(seconds)
    try:
        yield
    finally:
        signal.alarm(0)

# Use in generation
with timeout(30):  # 30 second timeout
    generated_ids = model.generate(...)
```

### 2. **Configure Axios Timeouts**
```javascript
// In server.js
const response = await axios.post(
    "http://localhost:5000/ask",
    { question: question.trim() },
    { timeout: 45000 }  // 45 second timeout
);
```

### 3. **Implement Retry Logic**
```javascript
// Using axios-retry
const axiosRetry = require('axios-retry');
axiosRetry(axios, {
    retries: 3,
    retryDelay: axiosRetry.exponentialDelay,
    retryCondition: (error) => {
        return error.code === 'ECONNABORTED' || 
               error.response?.status >= 500;
    }
});
```

### 4. **Frontend Timeout Handling**
```javascript
// In App.js askQuestion()
const controller = new AbortController();
const timeoutId = setTimeout(() => controller.abort(), 60000);

try {
    const res = await axios.post(`${API_BASE}/ask`, 
        { question: trimmedQuestion },
        { signal: controller.signal }
    );
} catch (e) {
    if (e.name === 'AbortError') {
        alert("Request timed out. Please try again.");
    }
} finally {
    clearTimeout(timeoutId);
}
```

## 📝 Implementation Steps
1. Install `axios-retry` package: `npm install axios-retry`
2. Add timeout to LLM generation in `main.py`
3. Configure axios timeout in `server.js`
4. Implement retry mechanism with exponential backoff
5. Add frontend timeout with clear error messages
6. Return 504 Gateway Timeout for timeout errors
7. Add logging for timeout events
8. Document timeout values in README

## 📂 Files to Modify
- `rag-service/main.py` - Add generation timeout
- `server.js` - Configure axios timeout & retry
- `frontend/src/App.js` - Frontend timeout handling
- `package.json` - Add axios-retry dependency
- `README.md` - Document timeout configuration

## 🧪 Testing Checklist
- [ ] Test with artificially slow model response
- [ ] Verify 504 error returned on timeout
- [ ] Test retry logic with transient failures
- [ ] Verify frontend shows timeout message
- [ ] Test concurrent requests don't block each other
- [ ] Verify successful requests complete normally
- [ ] Test exponential backoff timing
- [ ] Check logs contain timeout events

## 🎓 Acceptance Criteria
- [x] LLM generation has 30-second timeout
- [x] Axios requests have 45-second timeout
- [x] Retry mechanism with exponential backoff (3 attempts)
- [x] Frontend displays clear timeout message
- [x] 504 status code returned for timeouts
- [x] Hanging requests don't block service
- [x] Successful retry after transient failure
- [x] Timeout values configurable via environment variables

## ⚙️ Configuration (Environment Variables)
```bash
# .env
LLM_GENERATION_TIMEOUT=30  # seconds
API_REQUEST_TIMEOUT=45     # seconds
MAX_RETRY_ATTEMPTS=3
```

**Comment by Krish3017** [comment]
i would like to solve this issue , T152

**Comment by Aryan-B-Parikh** [comment]
I would like to work on this issue

Team 146
